File: nextjs-app/lib/services/rateLimitService.ts
/**
 * Сервис защиты от brute-force атак через rate limiting
 * Использует AuthAttemptsRepository для подсчёта попыток
 */

import type { AuthAttemptsRepository } from '@/lib/repositories/authAttemptsRepository';

/**
 * Результат проверки rate limit
 */
export interface RateLimitResult {
  /** Разрешена ли попытка входа */
  allowed: boolean;
  
  /** Количество попыток за период */
  attemptsCount: number;
  
  /** Время до разблокировки (если заблокирован) */
  blockedUntil: Date | null;
}

/**
 * Конфигурация rate limiting
 */
export interface RateLimitConfig {
  /** Максимальное количество попыток (по умолчанию 5) */
  maxAttempts?: number;
  
  /** Временное окно в минутах (по умолчанию 15) */
  windowMinutes?: number;
}

/**
 * RateLimitService - сервис rate limiting
 */
export class RateLimitService {
  private attemptsRepo: AuthAttemptsRepository;
  private maxAttempts: number;
  private windowMinutes: number;

  /**
   * Создаёт экземпляр RateLimitService
   * @param attemptsRepo - Репозиторий попыток входа
   * @param config - Конфигурация rate limiting
   */
  constructor(
    attemptsRepo: AuthAttemptsRepository,
    config?: RateLimitConfig
  ) {
    this.attemptsRepo = attemptsRepo;
    this.maxAttempts = config?.maxAttempts ?? 5;
    this.windowMinutes = config?.windowMinutes ?? 15;

    if (this.maxAttempts < 1) {
      throw new Error('maxAttempts must be >= 1');
    }

    if (this.windowMinutes < 1) {
      throw new Error('windowMinutes must be >= 1');
    }
  }

  /**
   * Проверяет rate limit для указанного tg_id
   * @param tgId - Telegram ID администратора
   * @returns Результат проверки rate limit
   */
  async checkRateLimit(tgId: number): Promise<RateLimitResult> {
    if (tgId < 1) {
      throw new Error('tgId must be >= 1');
    }

    try {
      // Подсчитываем попытки за последние N минут
      const attemptsCount = await this.attemptsRepo.countRecentAttempts(
        tgId,
        this.windowMinutes
      );

      // Если попыток меньше максимума - разрешаем
      if (attemptsCount < this.maxAttempts) {
        return {
          allowed: true,
          attemptsCount,
          blockedUntil: null,
        };
      }

      // Если попыток >= максимума - блокируем
      // Вычисляем время разблокировки (самая старая попытка + окно)
      const windowStart = new Date(
        Date.now() - this.windowMinutes * 60 * 1000
      );
      
      const oldestAttempt = await this.attemptsRepo.getOldestInWindow(
        tgId,
        windowStart
      );

      let blockedUntil: Date | null = null;
      
      if (oldestAttempt) {
        blockedUntil = new Date(
          oldestAttempt.timestamp.getTime() + this.windowMinutes * 60 * 1000
        );
      }

      return {
        allowed: false,
        attemptsCount,
        blockedUntil,
      };
    } catch (error) {
      console.error('Error checking rate limit:', error);
      throw new Error('Failed to check rate limit');
    }
  }

  /**
   * Записывает неудачную попытку входа
   * @param tgId - Telegram ID администратора
   * @param ipAddress - IP адрес (опционально)
   */
  async recordFailedAttempt(
    tgId: number,
    ipAddress: string | null = null
  ): Promise<void> {
    if (tgId < 1) {
      throw new Error('tgId must be >= 1');
    }

    try {
      await this.attemptsRepo.recordAttempt(tgId, ipAddress, false);
    } catch (error) {
      console.error('Error recording failed attempt:', error);
      throw new Error('Failed to record failed attempt');
    }
  }

  /**
   * Очищает все попытки после успешного входа
   * @param tgId - Telegram ID администратора
   */
  async clearAttempts(tgId: number): Promise<void> {
    if (tgId < 1) {
      throw new Error('tgId must be >= 1');
    }

    try {
      await this.attemptsRepo.clearAttempts(tgId);
    } catch (error) {
      console.error('Error clearing attempts:', error);
      throw new Error('Failed to clear attempts');
    }
  }

  /**
   * Получает количество оставшихся попыток
   * @param tgId - Telegram ID администратора
   * @returns Количество оставшихся попыток (0 если заблокирован)
   */
  async getRemainingAttempts(tgId: number): Promise<number> {
    if (tgId < 1) {
      throw new Error('tgId must be >= 1');
    }

    try {
      const attemptsCount = await this.attemptsRepo.countRecentAttempts(
        tgId,
        this.windowMinutes
      );

      const remaining = this.maxAttempts - attemptsCount;
      return remaining > 0 ? remaining : 0;
    } catch (error) {
      console.error('Error getting remaining attempts:', error);
      throw new Error('Failed to get remaining attempts');
    }
  }
}

/**
 * Создаёт экземпляр RateLimitService с конфигурацией из переменных окружения
 */
export function createRateLimitService(
  attemptsRepo: AuthAttemptsRepository
): RateLimitService {
  const maxAttempts = process.env.RATE_LIMIT_MAX_ATTEMPTS
    ? parseInt(process.env.RATE_LIMIT_MAX_ATTEMPTS, 10)
    : 5;

  const windowMinutes = process.env.RATE_LIMIT_WINDOW_MINUTES
    ? parseInt(process.env.RATE_LIMIT_WINDOW_MINUTES, 10)
    : 15;

  return new RateLimitService(attemptsRepo, {
    maxAttempts,
    windowMinutes,
  });
}

File: telegram-bot/services/rate_limit_service.py
"""
Сервис защиты от brute-force атак через rate limiting
"""

from dataclasses import dataclass
from datetime import datetime, timedelta, timezone
from typing import Optional
from utils.logger import get_logger

logger = get_logger(__name__)


@dataclass
class RateLimitResult:
    """
    Результат проверки rate limit
    
    Attributes:
        allowed: Разрешена ли попытка входа
        attempts_count: Количество попыток за окно времени
        blocked_until: Время до которого заблокирован (если заблокирован)
    """
    allowed: bool
    attempts_count: int
    blocked_until: Optional[datetime]


class RateLimitService:
    """
    Сервис rate limiting для защиты от brute-force атак
    
    Использует sliding window алгоритм:
    - Максимум 5 попыток за 15 минут
    - После 5-й неудачной попытки блокировка на 15 минут
    - Блокировка снимается автоматически через 15 минут
    
    Validates: Requirements 12.4, 12.5
    """
    
    def __init__(self, auth_attempts_repository):
        """
        Инициализирует сервис rate limiting
        
        Args:
            auth_attempts_repository: Репозиторий для работы с попытками входа
        """
        self._auth_attempts_repo = auth_attempts_repository
        self._max_attempts = 5
        self._window_minutes = 15
    
    async def check_rate_limit(self, tg_id: int) -> RateLimitResult:
        """
        Проверяет rate limit для администратора
        
        Логика:
        1. Подсчитывает попытки за последние 15 минут
        2. Если >= 5 попыток - блокирует
        3. Возвращает результат с информацией о блокировке
        
        Args:
            tg_id: Telegram ID администратора
        
        Returns:
            RateLimitResult с информацией о разрешении/блокировке
        
        Examples:
            >>> # После 4 попыток - разрешено
            >>> result = await service.check_rate_limit(123456789)
            >>> result.allowed
            True
            >>> result.attempts_count
            4
            
            >>> # После 5 попыток - заблокировано
            >>> result = await service.check_rate_limit(123456789)
            >>> result.allowed
            False
            >>> result.attempts_count
            5
            >>> result.blocked_until is not None
            True
        
        Validates: Requirements 12.4, 12.5
        """
        # Подсчитываем попытки за последние 15 минут
        attempts_count = await self._auth_attempts_repo.count_recent_attempts(
            tg_id=tg_id,
            minutes=self._window_minutes
        )
        
        # Проверяем превышение лимита
        if attempts_count >= self._max_attempts:
            # Получаем самую старую попытку в окне для расчёта времени разблокировки
            # ИСПРАВЛЕНИЕ: Используем datetime.now(timezone.utc) вместо utcnow()
            window_start = datetime.now(timezone.utc) - timedelta(minutes=self._window_minutes)
            oldest_attempt = await self._auth_attempts_repo.get_oldest_in_window(
                tg_id=tg_id,
                window_start=window_start
            )
            
            # Рассчитываем время разблокировки
            if oldest_attempt:
                blocked_until = oldest_attempt.timestamp + timedelta(minutes=self._window_minutes)
            else:
                # Если не нашли старую попытку, блокируем на 15 минут от текущего времени
                blocked_until = datetime.now(timezone.utc) + timedelta(minutes=self._window_minutes)
            
            logger.warning(
                "rate_limit_exceeded",
                extra={
                    "tg_id": tg_id,
                    "attempts_count": attempts_count,
                    "blocked_until": blocked_until.isoformat(),
                    "window_minutes": self._window_minutes
                }
            )
            
            return RateLimitResult(
                allowed=False,
                attempts_count=attempts_count,
                blocked_until=blocked_until
            )
        
        # Лимит не превышен - разрешаем
        return RateLimitResult(
            allowed=True,
            attempts_count=attempts_count,
            blocked_until=None
        )
    
    async def record_failed_attempt(self, tg_id: int, ip_address: Optional[str] = None) -> None:
        """
        Записывает неудачную попытку входа
        
        Args:
            tg_id: Telegram ID администратора
            ip_address: IP адрес (опционально)
        
        Examples:
            >>> await service.record_failed_attempt(123456789)
            >>> await service.record_failed_attempt(123456789, "192.168.1.1")
        
        Validates: Requirements 12.4, 12.5
        """
        await self._auth_attempts_repo.record_attempt(
            tg_id=tg_id,
            ip_address=ip_address
        )
        
        logger.info(
            "failed_attempt_recorded",
            extra={"tg_id": tg_id, "ip_address": ip_address}
        )
    
    async def clear_attempts(self, tg_id: int) -> None:
        """
        Очищает все попытки входа для администратора
        
        Вызывается после успешной аутентификации для сброса счётчика.
        
        Args:
            tg_id: Telegram ID администратора
        
        Examples:
            >>> # После успешного входа очищаем попытки
            >>> await service.clear_attempts(123456789)
        
        Validates: Requirements 12.4, 12.5
        """
        await self._auth_attempts_repo.clear_attempts(tg_id=tg_id)
